File: dp/평범한배낭.c
//백준 https://www.acmicpc.net/problem/12865
//2차원 배열 dp

#include <stdio.h>

int n;
int k;
int w[101];
int v[101];
int dp[100001][101] = {0,};

int main()
{
    int max = 0;

    freopen("../input.txt","r",stdin);
    scanf("%d %d",&n, &k);

    for(int i = 1; i<=n; i++)
    {
        scanf("%d %d",&w[i],&v[i]);
    }
    for(int i = 1; i<=k; i++)
    {
        for (int j = 1; j<=n;j++)
        {
        
            if(i-w[j] >= 0 && dp[i][j-1] < dp[i-w[j]][j-1] + v[j])
                dp[i][j] = dp[i-w[j]][j-1] + v[j];
            else
                dp[i][j] = dp[i][j-1];            
            if(dp[i][j] >max){
                max = dp[i][j];}
        }
    }
    printf("%d",max);
}

File: dp/평범한배낭.py
n, k = map(int,input().split())
v = [0]
w = [0]
for i in range(n):
	t1,t2 = map(int,input().split())
	w.append(t1)
	v.append(t2)
dp = [[0 for _ in range(n+1)] for _ in range(k+1)]

for i in range(k+1): # i == current_weight
	for j in range(n+1): # j == number of object
		if(w[j] <= i):
			dp[i][j] = max(dp[i][j-1],dp[i-w[j]][j-1] + v[j])
		else:
			dp[i][j] = dp[i][j-1]

print(dp[k][n])

# #배열 하나 사용
# n, k = map(int,input().split())
# v = [0]
# w = [0]
# for i in range(n):
# 	t1,t2 = map(int,input().split())
# 	w.append(t1)
# 	v.append(t2)

# dp = [0 for i in range(k+1)]
# for i in range(1,n+1): #물건
# 	for j in range(k,0,-1):# 무게
# 		if w[i] <= j:
# 			dp[j] = max(dp[j],dp[j-w[i]] + v[i])
# print(dp)



	